staging/src/k8s.io/apiserver/pkg/apis/apiserver/validation/validation.go

```diff
@@ -194,18 +216,75 @@ func validateClaimValidationRules(rules []api.ClaimValidationRule, fldPath *fiel
 func validateClaimMappings(m api.ClaimMappings, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
 
-	if len(m.Username.Claim) == 0 {
-		allErrs = append(allErrs, field.Required(fldPath.Child("username", "claim"), "claim name is required"))
+	if m.Username.Expression != nil {
+		if len(m.Username.Claim) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("username", "claim"), m.Username.Claim, "claim and expression can't both be set"))
+		}
+		if m.Username.Prefix != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("username", "prefix"), *m.Username.Prefix, "prefix can't be set when expression is set"))
+		}
+	} else {
+		if len(m.Username.Claim) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("username", "claim"), "claim name is required"))
+		}
+		if m.Username.Prefix == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("username", "prefix"), "prefix is required"))
+		}
 	}
-	// TODO(aramase): when Expression is added to PrefixedClaimOrExpression, check prefix and expression are not both set.
-	if m.Username.Prefix == nil {
-		allErrs = append(allErrs, field.Required(fldPath.Child("username", "prefix"), "prefix is required"))
+
+	if m.Groups.Expression != nil {
+		if len(m.Groups.Claim) > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("groups", "claim"), m.Groups.Claim, "claim and expression can't both be set"))
+		}
+		if m.Groups.Prefix != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("groups", "prefix"), *m.Groups.Prefix, "prefix can't be set when expression is set"))
+		}
+	} else {
+		if len(m.Groups.Claim) > 0 && m.Groups.Prefix == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("groups", "prefix"), "prefix is required when claim is set"))
+		}
+		if m.Groups.Prefix != nil && len(m.Groups.Claim) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("groups", "claim"), "non-empty claim name is required when prefix is set"))
+		}
 	}
-	if len(m.Groups.Claim) > 0 && m.Groups.Prefix == nil {
-		allErrs = append(allErrs, field.Required(fldPath.Child("groups", "prefix"), "prefix is required when claim is set"))
+
+	// TODO(aramase): should we make ClaimOrExpression.expression *string?
+	if len(m.UID.Claim) > 0 && len(m.UID.Expression) > 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("uid", "expression"), m.UID.Expression, "expression can't be set when claim is set"))
+	}
+
+	// There is no check for duplicate extra mapping keys. If multiple
+	// mappings have the same key, the result will be a concatenation of
+	// all the values with the order preserved.
+	for i, mapping := range m.Extra {
+		fldPath := fldPath.Child("extra").Index(i)
+		if len(mapping.Key) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("key"), "key is required"))
+		}
+		if len(mapping.ValueExpression) == 0 {
```

Comment:
Lets also validate the CEL expression by compiling it here (we cannot perform type checking because we don't know the schema of the JWT but we should always make sure that they return the right type such as string, bool, etc).  For the user info validation we should type check because we know the fields.  These should return the compiled CEL functions and different public function can choose to ignore those as return times.